### DISCUSSION TITLE
A public GitHub repo cannot use a preset package from a private GitHub repository, even if the Renovate app has access to both

### DISCUSSION THREAD

**Discussion by maxdanilov** [renovatebot/renovate#26300] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the problem

Hi, I believe this used to work in the past (at least one one of our repos) but stopped working around Oct 17th 2023.

If a GitHub repo's (let'a call it 1️⃣) Renovate config refers to a preset file stored in a private one (let's call it 2️⃣), e.g.:

```
// renovate.json5 from 1️⃣ 
  extends: [
    "github>myorg/my-private-config:default.json5"  // default.json5 is stored in 2️⃣ , i.e. in myorg/my-private-config
  ],
```

Then it works fine if 1️⃣  is private (provided the GitHub Renovate app has been granted permissions to access both 1️⃣  and 2️⃣) ; but does not work at all if 1️⃣ is public and 2️⃣  is private (and Renovate app has been granted permissions for both)

Here's how the Renovate log looks like from when 1️⃣ is public (❌):

```
DEBUG: GET https://api.github.com/repos/myorg/my-private-config/contents/default.json5 = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=108)
DEBUG: Preset file default.json5 not found in myorg/my-private-config
```

Here's how the Renovate log looks like when 1️⃣ is private (✅):

```
DEBUG: http statistics
    { 
      "https://api.github.com/repos/myorg/my-private-config/contents/default.json5 (GET,200)": 1,
```

It seems like this: if repo 1️⃣ is public, Renovate does not try to make API requests to GitHub using the Renovate app's credentials, getting 404s because 2️⃣ is private.

Expected behavior: as previously, it works for both when 1️⃣ is private or public.

Recreation repository does not make sense since one of them has two be private.

### Relevant debug logs

<details><summary>Logs</summary>

```

Copy/paste the relevant log(s) here, between the starting and ending backticks

```

</details>

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

**Comment by rarkins** [comment, marked as answer]
This is by design, for security reasons. It means that even if the GitHub token leaks somehow, it's limited to only the public repo. Otherwise a leaked token would give access to all your private repos. 

We plan to make this opt-out one day, although it's not in our near term plans

**Reply by maxdanilov** [reply]
I see, thank you for the explanation. Is there a way to configure Renovate to use a separate token for authenticating to GitHub API instead of not using a token at all? This way, I could make a scoped PAT just for this cause and still have Renovate work (leaking the PAT would only have a limited impact as I wouldn't grant it permissions to all the private repos)

**Reply by rarkins** [reply]
I think that might work if you set matchHost=https://api.github.com/repos/your-org/your-shared-config-repo/


The longer match should mean it takes precedence over the app token

**Reply by maxdanilov** [reply]
@rarkins I've tried changing the visibility of a repo to private (Renovate app has access to it), but in the Renovate logs for that repo it's still 404 to get the private preset file. Is there any additional caching for GitHup API responses on the Renovate side?

Renovate Web UI still displays the repo as public: 

<img width="571" alt="Screenshot 2023-12-20 at 16 22 32" src="https://github.com/renovatebot/renovate/assets/6655714/afea6b81-b071-4fe9-b704-cd7374fc77b3">

**Reply by rarkins** [reply]
Not that I can think of but maybe the backend hasn't picked up the change to private
